torch/_dynamo/variables/tensor.py

```diff
@@ -677,7 +682,59 @@ def redistribute_fn_with_prim_types(x):
             )
             if not self.source:
                 # Intermediary
-                unimplemented("Intermediary tensors with registered hooks - NYI")
+
+                # This function is only invoked if the passed-in func has no source,
+                # useful for intermediary created functions.
+                def _lifted_fn_src(name):
+                    base = name
+                    for i in itertools.count():
+                        if name not in tx.output.global_scope:
+                            src = GlobalSource(name)
+                            tx.output.guards.add(
+                                src.make_guard(GuardBuilder.WEAKREF_ALIVE)
+                            )
+                            tx.output.install_global(name, weakref.ref(fn))
+                            break
+                        name = f"{base}_{i}"
+                    return src
+
+                # Either we have a source already, or we must make a weakref to the original fn
+                # and reference that.
+                src = fn_var.source if fn_var.source else _lifted_fn_src(name)
+
+                from .builder import GraphArg
+
+                # This wraps our user provided fn with a function that intercedes and
+                # uses our `invoke` higher order op to record a hook invocation in bwd graph.
+                fn = functools.partial(trace_wrapped, fn=fn)
+
+                # This little piece of logic ensures we only ever lift a given hook up once
+                # this important, because a dynamo->aot_autograd invariant is no duplicate sources in inputs.
```

Comment:
@jansel remind me again why the hook needs to be an input to the Dynamo produced FX graph here? This seems to make the input/output convention from Dynamo more complicated